### ISSUE TITLE
non-article page view capture

### ISSUE THREAD

**Issue by lsh-0** [elifesciences/elife-metrics#60]
## Request
capture and serve page views on non-article pages, at current and historical locations.

## Proposal
* three new models: `PageType`, `Page` and `PageCount`

```
PageType:
    name, varchar

Page:
    type, foreign key to PageType
    canonical-name, varchar

PageCount:
    page, foreign key to Page
    views, integer
    date-recorded, datestamp
```

the current `Article->Metric` model won't support non-article counts because:
* `Metric` model is specialised to capture article views. it has `full`, `abstract`, `pdf` and `digest` fields for counts
* an 'Article' is a central concept in elife-metrics, literally everything depends on it 
* generalising the `Article` model to just `Page` would require a massive refactor

Further, we need to keep track of a page's location over time, so GA can be queried accordingly.

This could be done with another database table, to store a pattern that matches against URL paths and a start/end date pair, however:
* these patterns are not something that should be edited without strong change management
* may require post-processing logic for older, more complicated regular expressions
* will not change very often
* no benefits in being queried from a database above and beyond a simple data structure

### PageType
* `name` is either `blog-article`, `event`, `interview`, `labs-post` or `press-package` (at time of writing).
* `PageType` is populated manually
* Is connected to it's historical url patterns by it's `name`. 

### Page
* `canonical-name` is the unchanging name of the page. It may have lived at many different locations during it's existance, but it's always been page 'foobar' (or 'c637e02d')
* Is populated by the results returned by GA. Each result must normalise to the `canonical-name` or a new `Page` will be created
* Normalisation rules will differ by PageType (see elife_v1.py through elife_v4.py for Article normalisation)

### PageCount
* the number of views that page had at the given `path` on the given `date-recorded`
* Is populated by the results returned by GA after normalisation and aggregation (multiple results for the same canonical page but at different paths)

## history map

To query GA you need a collection of values, but the important ones are `filters` and `startDate` and `endDate`.

There will be a single map of:
`{page-type: [{filters: [string], start: datestamp, end: datestamp}]}` loaded at runtime.

For example: 
`{"interview": [{filters: ["ga:pagePath=~^/interview$", "ga:pagePath=~^/interview/.*"], start: "2017-01-01", end: None}]}`

The current era of URIs to populate this history map can be derived from the `api-raml/dist/api.raml` file, that describes the routes of these types:
https://github.com/elifesciences/api-raml/blob/develop/dist/api.raml

Older paths for these page types may be derived from the journal project's formula which contains an nginx map of old paths to new paths:
https://github.com/elifesciences/journal-formula/blob/master/salt/journal/config/etc-nginx-traits.d-redirect-existing-paths.conf

---

including @giorgiosironi and @thewilkybarkid for feedback

**Comment by lsh-0** [comment]
work is happening here: 
https://github.com/elifesciences/elife-metrics/tree/feat-non-article-views-capture-attempt2

I'm going to introduce historical epochs as a separate PR once this stuff is merged and stable, but am anticipating I'll be able to re-use some of the api/redirects/regex munging I was thrashing about with in the first attempt:
https://github.com/elifesciences/elife-metrics/tree/feat-non-article-views

**Comment by giorgiosironi** [comment]
> Metric model is specialised to capture article views. it has full, abstract, pdf and digest fields for counts

No problem with that, perhaps some of these article-specific models should be eventually renamed into `ArticleMetric` to better reflect this in an application that deals with more heterogeneous types

**Comment by giorgiosironi** [comment]
> canonical-name is the unchanging name of the page. It may have lived at many different locations during it's existance, but it's always been page 'foobar' (or 'c637e02d')

If this is a generic term for the various fields like `id` (`blog-article`) or `number` (`podcast-episode`), we usually go for `identifier`.

**Comment by giorgiosironi** [comment]
> Older paths for these page types may be derived from the journal project's formula which contains an nginx map of old paths to new paths

Don't know the requirements, but are we sure we should go back collecting data back to the beginning of time? What are the relevant Jira tickets?

**Comment by lsh-0** [comment]
ticket is here: https://elifesciences.atlassian.net/browse/ELPP-3196

this is my second attempt at this feature

I went with `name` instead of `canonical-name` in the model, so `page.name`. I could switch to `page.identifier` - it feels slightly more-right than `name`.
